spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -910,118 +911,143 @@ struct ProductionVisitor
-                                                            builder::equal(builder::id("i"), state().lahead_end)));
-                pushBuilder(ambiguous);
-                pb->parseError("ambiguous look-ahead token match", location);
-                popBuilder();
-
-                true_->addAssign(state().lahead, builder::integer(p.tokenID()));
-                true_->addAssign(state().lahead_end, builder::id("i"));
+                break;
             }
         }
 
         popBuilder();
     }
 
-    void getSyncProduction(const Production& p) {
-        std::set<Production> tokens;
+    // Get productions for lookaheads of a given production. If the passed
+    // production is non-terminal this function recurses until it finds
+    // lookaheads.
+    Result<std::set<Production>> getLookaheadProductions(const Production& p) {
+        // Validation.
+        if ( auto while_ = p.tryAs<production::While>(); while_ && while_->expression() )
+            return hilti::result::Error("&synchronized cannot be used on while loops with conditions");
 
+        std::set<Production> result;
+
+        // Bottom cases for recursion.
         if ( p.isLiteral() )
-            tokens.insert(p);
+            result.insert(p);
+        else if ( auto lahead = p.tryAs<production::LookAhead>() ) {
+            auto [alt1, alt2] = lahead->lookAheads();
+            for ( const auto& alt : {alt1, alt2} )
+                result.insert(alt.begin(), alt.end());
+        }
 
-        else if ( auto loop = p.tryAs<production::While>() )
-            tokens = std::get<1>(loop->lookAheadProduction().lookAheads());
+        // Otherwise recurse.
+        else if ( auto resolved = p.tryAs<production::Resolved>(); resolved || p.isNonTerminal() ) {
+            auto rhss = resolved ? grammar.resolved(*resolved).rhss() : p.rhss();
 
-        else if ( p.isNonTerminal() ) {
-            auto rhss = std::vector<std::vector<Production>>();
+            for ( const auto& rs : rhss )
+                for ( const auto& r : rs ) {
+                    auto tokens = getLookaheadProductions(r);
 
-            if ( const auto resolved = p.tryAs<production::Resolved>() )
-                rhss = grammar.resolved(*resolved).rhss();
-            else
-                rhss = p.rhss();
+                    if ( ! tokens )
+                        return tokens.error();
 
-            for ( const auto& rs : rhss )
-                for ( const auto& r : rs )
-                    if ( r.isLiteral() )
-                        tokens.insert(r);
+                    for ( const auto& token : *tokens ) {
+                        auto productions = getLookaheadProductions(token);
+
+                        if ( ! productions )
+                            return hilti::result::Error(productions.error());
+
+                        for ( auto p : *productions )
+                            result.insert(p);
+                    }
+                }
         }
+        return result;
```

Comment:
This block isn't working yet I think. Couple issues:

1. The iteration shouldn't be limited to `Resolved` productions, but need to operate on all productions that have rhss. 
2. It needs to find the *1st* literal in `rs`, currently it's using all.

Finding the 1st one is actually a bit tricky because it needs to take the semantics of productions into account (like if it's a epsilon production or nullable). I have a different idea here, don't recall if we discussed this already: we already have code that's very similar in the `grammar.c`: https://github.com/zeek/spicy/blob/main/spicy/toolchain/src/compiler/codegen/grammar.cc#L279-L316
I think we should be able to factor that out that logic there and give `Grammar` a new method `vector<Production> getLookAheadLiterals(const Production& p)` that you can just call in `syncProduction()`. Let me know if you want me to give that refactoring a try.

Reply:
I pulled in your changes from https://github.com/zeek/spicy/tree/topic/robin%2Ferror_recovery_lah_refactor.